crates/algos/src/dss.rs

```diff
@@ -0,0 +1,203 @@
+            // is the smaller value, we need to swap ids so we update
+            // only the value for id2, not id1.
+            if id1 < id2 {
+                let tmp = id2;
+                id2 = id1;
+                id1 = tmp;
+            }
+
+            let old_entry = id1;
+            let new_entry = id2;
+
+            if self.update_parent(id1, old_entry, new_entry).is_ok() {
+                break;
+            }
+        }
+    }
+
+    /// Find the set of `id`.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// use graph::dss::DisjointSetStruct;
+    ///
+    /// let dss = DisjointSetStruct::new(10);
+    /// assert_eq!(dss.find(4), 4);
+    /// dss.union(4, 2);
+    /// assert_eq!(dss.find(4), 2);
+    /// ```
+    pub fn find(&self, mut id: NI) -> NI {
+        let mut parent = self.parent(id);
+
+        while id != parent {
+            let grand_parent = self.parent(parent);
+            // Try to apply path-halving by setting the value
+            // for some id to its grand parent. This might fail
+            // if another thread is also changing the same value
+            // but that's ok. The CAS operations guarantees
+            // that at least one of the contenting threads will
+            // succeed. That's enough for the path-halving to work
+            // and there is no need to retry in case of a CAS failure.
+            self.update_parent(id, parent, grand_parent).unwrap();
+            id = parent;
+            parent = grand_parent;
+        }
+
+        id
+    }
+
+    #[inline]
+    fn parent(&self, i: NI) -> NI {
+        self.0[i.index()].load(Ordering::Acquire)
+    }
+
+    #[inline]
+    fn update_parent(&self, id: NI, current: NI, new: NI) -> Result<NI, NI> {
+        self.0[id.index()].compare_exchange_weak(current, new, Ordering::Release, Ordering::Relaxed)
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use std::sync::Arc;
+
+    use super::*;
+
+    #[test]
+    fn test_union() {
+        let dss = DisjointSetStruct::new(10);
+
+        assert_eq!(dss.find(9), 9);
+        dss.union(9, 7);
+        assert_eq!(dss.find(9), 7);
+        dss.union(7, 4);
+        assert_eq!(dss.find(9), 4);
+        dss.union(4, 2);
+        assert_eq!(dss.find(9), 2);
+        dss.union(2, 0);
+        assert_eq!(dss.find(9), 0);
+    }
+
+    #[test]
+    fn test_union_parallel() {
+        let dss = Arc::new(DisjointSetStruct::new(1000));
+
+        let dss1 = dss.clone();
+        let dss2 = dss.clone();
+
+        let t1 = std::thread::spawn(move || {
+            for i in 0..500 {
+                dss1.union(i, i + 1);
+            }
+            for i in 501..999 {
+                dss1.union(i, i + 1);
+            }
+        });
+
+        let t2 = std::thread::spawn(move || {
+            for i in 0..500 {
+                dss2.union(i, i + 1);
+            }
+            for i in 501..999 {
+                dss2.union(i, i + 1);
+            }
+        });
```

Comment:
probably want to use a [Barrier](https://doc.rust-lang.org/stable/std/sync/struct.Barrier.html) to make sure that the computations actually run in parallel.

Here is a suggestion with some other small refactorings

```suggestion
        use std::sync::Barrier;

        let barrier = Arc::new(Barrier::new(2));
        let dss = Arc::new(DisjointSetStruct::new(1000));

        fn workload(barrier: &Barrier, dss: &DisjointSetStruct<u64>) {
            barrier.wait();

            for i in 0..500 {
                dss.union(i, i + 1);
            }

            // We wait again after the first cluster to increase the chance of concurrent updates
            barrier.wait();

            for i in 501..999 {
                dss.union(i, i + 1);
            }
        }

        let t1 = std::thread::spawn({
            let barrier = Arc::clone(&barrier);
            let dss = Arc::clone(&dss);

            move || workload(&barrier, &dss)
        });

        let t2 = std::thread::spawn({
            let barrier = Arc::clone(&barrier);
            let dss = Arc::clone(&dss);

            move || workload(&barrier, &dss)
        });
```